src/index.js

```diff
@@ -269,36 +279,41 @@ export default function undoable (reducer, rawConfig = {}) {
         }
 
         if (config.filter && typeof config.filter === 'function') {
-          if (!config.filter(action, res, state && state.present)) {
+          if (!config.filter(action, res, history.present)) {
             debug('filter prevented action, not storing it')
             debugEnd()
             return {
-              ...state,
+              ...history,
               present: res
             }
           }
         }
 
-        let updatedHistory
-        if (!isHistory(state)) {
-          updatedHistory = createHistory(state)
-          debug('create history on init')
-        } else if (state.present === res) {
-          updatedHistory = state
-          debug('not inserted, state is unchanged')
-        } else {
-          updatedHistory = insert(state, res, config.limit)
+        if (history.present !== res) {
+          history = insert(history, res, config.limit)
           debug('inserted new state into history')
+        } else {
+          debug('not inserted, history is unchanged')
         }
```

Comment:
No more need to create a history here. And yes, this could be simplified if think we don't need to call `debug('not inserted, history is unchanged')` here.